core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1013 @@
+	type Output = Self;
+
+	fn add(self, rhs: Self) -> Self::Output {
+		Self(self.0 + rhs.0)
+	}
+}
+
+/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
+/// for safe subtraction.
+impl ops::Sub for Fixed64 {
+	type Output = Self;
+
+	fn sub(self, rhs: Self) -> Self::Output {
+		Self(self.0 - rhs.0)
+	}
+}
+
+impl CheckedSub for Fixed64 {
+	fn checked_sub(&self, rhs: &Self) -> Option<Self> {
+		if let Some(v) = self.0.checked_sub(rhs.0) {
+			Some(Self(v))
+		} else {
+			None
+		}
+	}
+}
+
+impl CheckedAdd for Fixed64 {
+	fn checked_add(&self, rhs: &Self) -> Option<Self> {
+		if let Some(v) = self.0.checked_add(rhs.0) {
+			Some(Self(v))
+		} else {
+			None
+		}
+	}
+}
+
+/// Helper gcd function used in Rational128 implementation.
+fn gcd(a: u128, b: u128) -> u128 {
+	match ((a, b), (a & 1, b & 1)) {
+		((x, y), _) if x == y => y,
+		((0, x), _) | ((x, 0), _) => x,
+		((x, y), (0, 1)) | ((y, x), (1, 0)) => gcd(x >> 1, y),
+		((x, y), (0, 0)) => gcd(x >> 1, y >> 1) << 1,
+		((x, y), (1, 1)) => {
+			let (x, y) = (min(x, y), max(x, y));
+			gcd((y - x) >> 1, x)
+		},
+		_ => unreachable!(),
+    }
+}
+
+/// A wrapper for any rational number with a 128 bit numerator and denominator.
+#[derive(Clone, Copy, Default)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Rational128(u128, u128);
+
+/// Some helper functions to work with 128bit numbers. Note that the functionality provided here is
+/// only sensible to use with 128bit numbers because for smaller sizes, you can always rely on
+/// assumptions of a bigger type (u128) being available, or simply create a per-thing and use the
+/// multiplication implementation provided there.
+pub mod helpers_128bit {
+	/// return the number of bits that are needed to store `n`.
+	pub(crate) fn bits_of(n: u128) -> u32 {
+		128 - n.leading_zeros()
+	}
+
+	/// Tries to compute `a * b / c`. The approach is:
+	///   - Simply try a * b / c.
+	///   - Swap the operations in case the former multiplication overflows. Divide first. This
+	///     might collapse to zero.
+	///
+	/// If none worked then return Error.
+	pub fn multiply_by_rational(a: u128, b: u128, c: u128) -> Result<u128, &'static str> {
+		// This is the safest way to go. Try it.
+		if let Some(x) = a.checked_mul(b) {
+			Ok(x / c)
+		} else {
+			let bigger = a.max(b);
+			let smaller = a.min(b);
+			if bigger < c { return Err("division will collapse to zero"); }
+			(bigger / c).checked_mul(smaller).ok_or("multiplication overflow")
+		}
+	}
+
+	/// Performs [`multiply_by_rational`]. In case of failure, it greedily tries to shift the
+	/// numerator (`b`) and denominator (`c`) (looking at the whole thing is `a * b / c`) until the
+	/// multiplication fits. This is guaranteed to work if `b < c`.
+	///
+	/// In case `b > c` and overflow happens, `a` is returned.
+	///
+	/// c must be greater than or equal to 1.
+	pub fn safe_multiply_by_rational(a: u128, b: u128, c: u128) -> u128 {
```

Comment:
not sure safe_ is a good name maybe multiply_by_rational_or_$shortNameOfTheAlgo

Reply:
well, that's one of my problems here. I don't know the name of the algorithm and I cam up with it. I think it is simple and sound, but a proof check would be great. 

As for the name, `safe_` implies that it always returns, but it actually can return an inaccurate value. I also like the name `multiply_by_rational_or_greedy()`?  The initial version actually attempted to greedily shift to find the spot where the multiplication works. Recently I changed it to something a bit smarter.